usr/src/uts/common/fs/zfs/metaslab.c

```diff
@@ -2854,9 +2933,11 @@ metaslab_group_alloc_normal(metaslab_group_t *mg, zio_alloc_list_t *zal,
 		 * We have just failed an allocation attempt, check
 		 * that metaslab_should_allocate() agrees. Otherwise,
 		 * we may end up in an infinite loop retrying the same
-		 * metaslab.
+		 * metaslab. Keep in mind, it might have happened due
+		 * to an ongoing trim.
```

Comment:
According to the comment, the concern is not about repeatedly trying all metaslabs, it's about retrying this one metaslab over and over even though we can't allocate from it.  The loop looks for the "best" metaslab each time.  If this metaslab seems like the "best" (based on its weight), but actually we can't allocate from it because of in-progress trims, we may retry this same metaslab again.